### ISSUE TITLE
Naga generates non-functional SPIR-V from WGSL

### ISSUE THREAD

**Issue by Kneelawk** [gfx-rs/naga#987]
## The Issue
I have a project where I generate a SPIR-V shader from a WGSL source. Any
SPIR-V binary that Naga generates from my WGSL source file is not functional
in my WGPU render pipeline, regardless of whether Naga is being used as a
standalone CLI or as a library. However, using Tint to translate the my WGSL
source file into SPIR-V does produce a functional SPIR-V binary. Using my WGSL
shader directly in the WGPU render pipeline also works.

Using Naga to translate the WGSL source to SPIR-V, either through the CLI or as a
library, produces a shader that, when run in this render pipeline, produces a
blank image that is the same color as the clear color. Non-Naga shader sources,
when run in this render pipeline, produce a blue triangle in the center of the
image.

## Replicating The Issue
My code for replicating this issue is at https://github.com/Kneelawk/naga-error

## System Information
 * OS: Linux Kubuntu 20.04.2 LTS
 * Rust Version: rustc 1.55.0-nightly (f586d79d1 2021-06-13)
 * Naga Version: master#efd416d96448d0c749d332b857b67bed3b7fc66e
 * WGPU Version: 0.8.1
 * Tint Version: main#52b6a004b8b74ca2750b927bd020fdd8352db4fa
 * Vulkan Instance Version: 1.2.131
 * GPU String: GeForce GTX 1060 6GB

**Comment by kvark** [comment]
This is a well-prepared issue. It's highly appreciated, thank you!
There is nothing wrong with the generated SPIR-V. In fact, Naga's output is conceptually simpler than Tint's. We could use better debug labels though.

Generally when looking at graphics bugs, it helps to know the graphics debugging tools. I used XCode for this.
Here is what I did:
  1. inserted `device.start_capture()` and `device.stop_capture()` around the logic
  2. created a new XCode project with "external build system" (which it picks to be Cargo for me, somehow).
  3. launched from XCode. The GPU capture is automatically triggered because of (1)
  4. looked at the draw call state. No polygons were generated, which means only that the generated polygons where backface culled.

Now, the reason why everything is backface culled with Naga but not Tint lies in the coordinate spaces. I don't know what convention Tint does when converting to/from SPIR-V, but Naga's binary is configured by default to consider each shader in its own "native" coordinate system. That makes no difference for everybody, except SPIR-V. So when we see SPIR-V on the input, in naga CLI binary, we treat it as Y-flipped. Same for producing SPIR-V.

However, when SPIR-V is fed to wgpu, we expect it to be given in WebGPU coordinate system. So we disable any flipping logic in Naga. Your code works as expected:
  - If you load the WGSL, it displays correctly, and it's written in the proper space for the program.
  - If you convert it with Tint to SPIR-V, looks like Tint doesn't do any coordinate space adjustment by default. So the produced SPIR-V is still made for the same coordinate system, and we take it as such, so it renders.
  - If you convert it with Naga to SPIR-V, and you use the default options to do so, Naga will produce the coordinate system transformation into this SPIR-V. So you could take it and run with Vulkan straight (since Naga is made for more than just WebGPU needs). But when you load it in wgpu, it assumes WebGPU coordinate system, and your polygons are culled.

In other words, wgpu-naga workflows do not assume that you are going to be feeding the produced SPIRV back. You'd instead either produce WGSL, or at least configured the SPIR-V backend to skip the coordinate space adjustment:
```rust
let options = back::spv::Options {
        flags: back::spv::WriterFlags::empty(), // no `ADJUST_COORDINATE_SPACE` flag
        ..Default::default()
    };
```

Alternatively, you can see what's going on by either changing the `front_face` or disabling the `cull_mode` on the pipeline.
